src/CodeGen/GenerateSyntax.hs

```diff
@@ -13,32 +13,62 @@ module CodeGen.GenerateSyntax
 
 import Data.Char
 import Language.Haskell.TH
+import Data.HashSet (HashSet)
 import Language.Haskell.TH.Syntax as TH
 import CodeGen.Deserialize (MkDatatype (..), MkDatatypeName (..), MkField (..), MkRequired (..), MkType (..), MkNamed (..), MkMultiple (..))
 import Data.List.NonEmpty (NonEmpty (..))
 import Data.Foldable
 import Data.Text (Text)
 import qualified Data.HashSet as HashSet
-import Data.HashSet (HashSet)
+import qualified TreeSitter.Importing as TS
+import GHC.Generics hiding (Constructor, Datatype)
+import Foreign.Ptr
+import qualified TreeSitter.Language as TS
+import Foreign.C.String
+import Data.Proxy
 
 -- Auto-generate Haskell datatypes for sums, products and leaf types
-datatypeForConstructors :: MkDatatype -> Q Dec
-datatypeForConstructors (SumType (DatatypeName datatypeName) named subtypes) = do
+datatypeForConstructors :: Ptr TS.Language -> MkDatatype -> Q [Dec]
+datatypeForConstructors language (SumType (DatatypeName datatypeName) named subtypes) = do
   let name = toName' named datatypeName
   cons <- traverse (toSumCon datatypeName) subtypes
-  pure $ DataD [] name [] Nothing cons [ DerivClause Nothing [ ConT ''Eq, ConT ''Ord, ConT ''Show ] ]
-datatypeForConstructors (ProductType (DatatypeName datatypeName) named fields) = do
+  result <- symbolMatchingInstanceForSums language name subtypes
+  pure $ DataD [] name [] Nothing cons [ DerivClause Nothing [ ConT ''TS.Building, ConT ''Eq, ConT ''Generic, ConT ''Ord, ConT ''Show ] ]:result
```

Comment:
Can we reduce the duplication in these stanzas? I would probably start with pulling out the `[ConT ''TS.Building, etc.]` into

```
standardASTDerivations :: DerivClause
standardASTDerivations = DerivClause Nothing [ ConT ''TS.Building, ConT ''Eq, ConT ''Generic, ConT ''Ord, ConT ''Show ] ]
```

But I also wonder (and @robrix might know here)—can we use the literal `[d| … ]` quasiquotation style here? It is significantly more readable, though I don't know if there are any technical reasons why we can't.

Reply:
Unfortunately not; AFAIK, `[d|…|]` won’t let us splice in the datatype’s name.

We could maybe simplify a little by using the smart constructors, tho, e.g. `dataD` & friends.

Reply:
@patrickt I went with the longhand because I couldn't find official TH documentation on the quasi quotation style, so this was easier to debug and find references in Dash for. I agree `[d|...]` is more readable and shorter but if I have to come back to this code, I will rely on Dash more than my memory. 

I should def factor out the duplication though.